stan/math/prim/fun/gaus_interp.hpp

```diff
@@ -0,0 +1,196 @@
+  auto lb = std::lower_bound(xs.begin(), xs.end(), x - NSTDS * sig);
+  int ind_start = &(*lb) - &xs[0] - 1;
+  ind_start = std::max(0, ind_start);
+
+  auto ub = std::upper_bound(xs.begin(), xs.end(), x + NSTDS * sig);
+  int ind_end = &(*ub) - &xs[0];
+  ind_end = std::min(n - 1, ind_end);
+
+  // sum convolutions over intervals
+  return_type_t<Tx> y = 0;
+  for (int i = ind_start; i < ind_end; i++) {
+    y += conv_gaus_line(xs2[i], xs2[i + 1], params[i], params[(n - 1) + i], x,
+                        params[2 * n - 2]);
+  }
+  return y;
+}
+
+/**
+ * This function was written to be used with gaus_interp. This function
+ * computes the shifted y-values of the reference points of an interpolation
+ * in such a way that when that piecewise linear function is convolved
+ * with a Gaussian kernel, the resulting function coincides with the
+ * points \f$(xs_i, ys_i)\f$ inputted into this function. The output of this
+ * function depends heavily on the choice of width of the Gaussian
+ * kernel, which at the time of writing, is set to one tenth the
+ * minimum distance between successive elements of the vector xs.
+ * A tolerance for the maximum distance between the interpolation and
+ * all reference points is also set manually and is not an input.
+ *
+ *
+ * @param xs vector of independent variable of reference points
+ * @param ys vector of dependent variable of reference points
+ * @return struct containing slopes, intercepts, and width of kernel
+ */
+inline std::vector<double> gaus_interp_precomp(const std::vector<double>& xs,
+                                               const std::vector<double>& ys) {
+  static char const* function = "gaus_interp_precomp";
+  check_not_nan(function, "xs", xs);
+  check_not_nan(function, "ys", ys);
+  check_ordered(function, "xs", xs);
+  check_greater(function, "xs", xs.size(), 1);
+
+  using internal::min_diff;
+  static const double max_diff = 1e-8;
+  // set Gaussian kernel to sig2_scale times smallest difference between
+  // successive points
+  static const double sig2_scale = 0.1;
+  int n = xs.size();
+
+  // create the vector to be returned that consists of as, bs, sig2
+  std::vector<double> params;
+  params.resize(2 * n - 1);
+  params[2 * n - 2] = square(min_diff(n, xs) * sig2_scale);
+
+  // copy ys into a new vector that will be changed
+  std::vector<double> y2s = ys;
+
+  // interatively find interpolation that coincides with ys at xs
+  int max_iters = 50;
+  double dd;
+  for (int j = 0; j < max_iters; j++) {
+    // find slope and intercept of line between each point
+    for (size_t i = 0; i < n - 1; i++) {
+      params[i] = (y2s[i + 1] - y2s[i]) / (xs[i + 1] - xs[i]);
+      params[(n - 1) + i] = -xs[i] * params[i] + y2s[i];
+    }
+
+    double dmax = 0;
+    for (size_t i = 0; i < n; i++) {
+      dd = ys[i] - gaus_interp(xs, y2s, params, xs[i]);
+      y2s[i] += dd;
+      dmax = std::max(std::abs(dd), dmax);
+    }
+    if (dmax < max_diff)
+      break;
```

Comment:
[optional] I prefer having scopes defined even for one liners

```suggestion
    if (dmax < max_diff) {
      break;
    }
```